Panic on next command

```
(dlv) trace
Command failed: Malformed breakpoint location "" at 0: empty string
(dlv) trace math.go:10
Command failed: Breakpoint exists at /usr/src/go/src/github.com/krhubert/optimization/math.go:10 at 40103d
(dlv) trace math.go:11
Command failed: no code at /usr/src/go/src/github.com/krhubert/optimization/math.go:11
(dlv) trace math.go:9
Tracepoint 3 set at 0x40102b for main.main ./math.go:9
(dlv) c
> main.main() ./math.go:8 (hits goroutine(1):1 total:1)
     3:	import "math"
     4:	
     5:	var f = 1.5
     6:	
     7:	func main() {
=>   8:		_ = math.Floor(f)
     9:		_ = float64(int(f))
    10:	}
(dlv) print f
1.5
(dlv) n
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5110f6]

goroutine 1 [running]:
github.com/derekparker/delve/terminal.printcontextThread(0xc8204ae820, 0xc820461a90)
	/usr/src/go/src/github.com/derekparker/delve/terminal/command.go:892 +0x13b6
github.com/derekparker/delve/terminal.printcontext(0xc8204ae820, 0xc8204f43c0, 0x0, 0x0)
	/usr/src/go/src/github.com/derekparker/delve/terminal/command.go:873 +0x34d
github.com/derekparker/delve/terminal.next(0xc8204ae820, 0x0, 0x0, 0x0, 0x0)
	/usr/src/go/src/github.com/derekparker/delve/terminal/command.go:489 +0x8d
github.com/derekparker/delve/terminal.(*Term).Run(0xc8204ae820, 0x0, 0x0, 0x0)
	/usr/src/go/src/github.com/derekparker/delve/terminal/terminal.go:117 +0xab2
main.execute(0x0, 0xc82010a610, 0x1, 0x1, 0xc82010c060, 0x0)
	/usr/src/go/src/github.com/derekparker/delve/cmd/dlv/main.go:362 +0x807
main.init.1.func4(0xc82011c800, 0xc82010a610, 0x1, 0x1)
	/usr/src/go/src/github.com/derekparker/delve/cmd/dlv/main.go:119 +0x4a
github.com/spf13/cobra.(*Command).execute(0xc82011c800, 0xc82010a5a0, 0x1, 0x1, 0x0, 0x0)
	/usr/src/go/src/github.com/spf13/cobra/command.go:569 +0x869
github.com/spf13/cobra.(*Command).ExecuteC(0xc82011c000, 0xc82011c800, 0x0, 0x0)
	/usr/src/go/src/github.com/spf13/cobra/command.go:656 +0x56b
github.com/spf13/cobra.(*Command).Execute(0xc82011c000, 0x0, 0x0)
	/usr/src/go/src/github.com/spf13/cobra/command.go:615 +0x2d
main.main()
	/usr/src/go/src/github.com/derekparker/delve/cmd/dlv/main.go:186 +0x3c

goroutine 17 [syscall, 11 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 5 [syscall, 11 minutes]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 7 [IO wait]:
net.runtime_pollWait(0x7ff73f0ec300, 0x72, 0xc82000c1b0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82011e610, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82011e610, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82011e5b0, 0xc820a64001, 0x1ffdff, 0x1ffdff, 0x0, 0x7ff741f18050, 0xc82000c1b0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820026090, 0xc820a64001, 0x1ffdff, 0x1ffdff, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
encoding/json.(*Decoder).refill(0xc82005da00, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:152 +0x287
encoding/json.(*Decoder).readValue(0xc82005da00, 0x1, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:128 +0x41b
encoding/json.(*Decoder).Decode(0xc82005da00, 0x9669e0, 0xc8200864c8, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:57 +0x159
net/rpc/jsonrpc.(*clientCodec).ReadResponseHeader(0xc820086480, 0xc82022dd40, 0x0, 0x0)
	/usr/local/go/src/net/rpc/jsonrpc/client.go:75 +0x8f
net/rpc.(*Client).input(0xc8204c6120)
	/usr/local/go/src/net/rpc/client.go:109 +0xbf
created by net/rpc.NewClientWithCodec
	/usr/local/go/src/net/rpc/client.go:201 +0xd2

goroutine 8 [select, locked to thread]:
runtime.gopark(0xb502e0, 0xc82001e728, 0xa4f0b0, 0x6, 0x431118, 0x2)
	/usr/local/go/src/runtime/proc.go:185 +0x163
runtime.selectgoImpl(0xc82001e728, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:392 +0xa64
runtime.selectgo(0xc82001e728)
	/usr/local/go/src/runtime/select.go:212 +0x12
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:227 +0x353
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 23 [IO wait]:
net.runtime_pollWait(0x7ff73f0ec240, 0x72, 0xc82000c1b0)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82011e680, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82011e680, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82011e620, 0xc820123201, 0x5ff, 0x5ff, 0x0, 0x7ff741f18050, 0xc82000c1b0)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc82010c0e0, 0xc820123201, 0x5ff, 0x5ff, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
encoding/json.(*Decoder).refill(0xc8201b04e0, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:152 +0x287
encoding/json.(*Decoder).readValue(0xc8201b04e0, 0x1, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:128 +0x41b
encoding/json.(*Decoder).Decode(0xc8201b04e0, 0x966aa0, 0xc820117340, 0x0, 0x0)
	/usr/local/go/src/encoding/json/stream.go:57 +0x159
net/rpc/jsonrpc.(*serverCodec).ReadRequestHeader(0xc820117320, 0xc820227fe0, 0x0, 0x0)
	/usr/local/go/src/net/rpc/jsonrpc/server.go:66 +0x7c
net/rpc.(*Server).readRequestHeader(0xc82010f980, 0x7ff73f0ec8c0, 0xc820117320, 0x0, 0x0, 0xc820227fe0, 0xc820043c00, 0x0, 0x0)
	/usr/local/go/src/net/rpc/server.go:576 +0x90
net/rpc.(*Server).readRequest(0xc82010f980, 0x7ff73f0ec8c0, 0xc820117320, 0xc82010f980, 0xc8201af0b0, 0xc820110800, 0x0, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/net/rpc/server.go:543 +0x8b
net/rpc.(*Server).ServeCodec(0xc82010f980, 0x7ff73f0ec8c0, 0xc820117320)
	/usr/local/go/src/net/rpc/server.go:462 +0x8c
github.com/derekparker/delve/service/rpc.(*ServerImpl).Run.func1(0xc82010c0a0)
	/usr/src/go/src/github.com/derekparker/delve/service/rpc/server.go:72 +0x4f9
created by github.com/derekparker/delve/service/rpc.(*ServerImpl).Run
	/usr/src/go/src/github.com/derekparker/delve/service/rpc/server.go:73 +0x122

goroutine 9 [chan receive, 11 minutes]:
github.com/derekparker/delve/terminal.(*Term).Run.func1(0xc8204c6360, 0xc8204ae820)
	/usr/src/go/src/github.com/derekparker/delve/terminal/terminal.go:56 +0x77
created by github.com/derekparker/delve/terminal.(*Term).Run
	/usr/src/go/src/github.com/derekparker/delve/terminal/terminal.go:62 +0x1c2

goroutine 102 [chan receive, locked to thread]:
github.com/derekparker/delve/proc.(*Process).handlePtraceFuncs(0xc820092000)
	/usr/src/go/src/github.com/derekparker/delve/proc/proc.go:783 +0xc3
created by github.com/derekparker/delve/proc.New
	/usr/src/go/src/github.com/derekparker/delve/proc/proc.go:78 +0x1d4
SIGTRAP: trace trap
PC=0x40102c m=0

goroutine 1 [running]:
main.main()
	/usr/src/go/src/github.com/krhubert/optimization/math.go:9 +0x2c fp=0xc820039f50 sp=0xc820039f38
runtime.main()
	/usr/local/go/src/runtime/proc.go:111 +0x2b0 fp=0xc820039fa0 sp=0xc820039f50
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1 fp=0xc820039fa8 sp=0xc820039fa0

rax    0x1
rbx    0x4b26c0
rcx    0x432fffffffffffff
rdx    0x3ff7ffffffffffff
rdi    0xc8200120d0
rsi    0x7
rbp    0x0
rsp    0xc820039f38
r8     0x457cc0
r9     0xc820000f00
r10    0x8
r11    0x206
r12    0x426470
r13    0x486d14
r14    0x0
r15    0x8
rip    0x40102c
rflags 0x287
cs     0x33
fs     0x0
gs     0x0
```

--- a/_fixtures/math.go
+++ b/_fixtures/math.go
@@ -1,0 +1,10 @@
+package main
+
+import "math"
+
+var f = 1.5
+
+func main() {
+	_ = math.Floor(f)
+	_ = float64(int(f))
+}

--- a/terminal/command.go
+++ b/terminal/command.go
@@ -940,7 +940,7 @@ func printcontext(t *Term, state *api.DebuggerState) error {
 
 	printcontextThread(t, state.CurrentThread)
 
-	if state.CurrentThread.Breakpoint == nil || !state.CurrentThread.Breakpoint.Tracepoint {
+	if state.CurrentThread.Breakpoint == nil || !state.CurrentThread.Breakpoint.Tracepoint || state.CurrentThread.BreakpointInfo == nil {
 		return printfile(t, state.CurrentThread.File, state.CurrentThread.Line, true)
 	}
 	return nil
@@ -955,7 +955,7 @@ func printcontextThread(t *Term, th *api.Thread) {
 	}
 
 	args := ""
-	if th.Breakpoint.Tracepoint {
+	if th.Breakpoint.Tracepoint && th.BreakpointInfo != nil {
 		var arg []string
 		for _, ar := range th.BreakpointInfo.Arguments {
 			arg = append(arg, ar.SinglelineString())

--- a/terminal/command_test.go
+++ b/terminal/command_test.go
@@ -2,12 +2,78 @@ package terminal
 
 import (
 	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
 	"path/filepath"
+	"strings"
 	"testing"
 
 	"github.com/derekparker/delve/proc/test"
+	"github.com/derekparker/delve/service"
 	"github.com/derekparker/delve/service/api"
+	"github.com/derekparker/delve/service/rpc"
 )
+
+type FakeTerminal struct {
+	t      testing.TB
+	client service.Client
+	cmds   *Commands
+	term   *Term
+}
+
+func (term *FakeTerminal) Exec(cmdstr string) (outstr string, err error) {
+	cmdstr, args := parseCommand(cmdstr)
+	cmd := term.cmds.Find(cmdstr)
+
+	outfh, err := ioutil.TempFile("", "cmdtestout")
+	if err != nil {
+		term.t.Fatalf("could not create temporary file: %v", err)
+	}
+
+	stdout, stderr := os.Stdout, os.Stderr
+	os.Stdout, os.Stderr = outfh, outfh
+	defer func() {
+		os.Stdout, os.Stderr = stdout, stderr
+		outfh.Close()
+		outbs, err1 := ioutil.ReadFile(outfh.Name())
+		if err1 != nil {
+			term.t.Fatalf("could not read temporary output file: %v", err)
+		}
+		outstr = string(outbs)
+		os.Remove(outfh.Name())
+	}()
+	err = cmd(term.term, args)
+	return
+}
+
+func (term *FakeTerminal) MustExec(cmdstr string) string {
+	outstr, err := term.Exec(cmdstr)
+	if err != nil {
+		term.t.Fatalf("Error executing <%s>: %v", cmdstr, err)
+	}
+	return outstr
+}
+
+func withTestTerminal(name string, t testing.TB, fn func(*FakeTerminal)) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("couldn't start listener: %s\n", err)
+	}
+	defer listener.Close()
+	server := rpc.NewServer(&service.Config{
+		Listener:    listener,
+		ProcessArgs: []string{test.BuildFixture(name).Path},
+	}, false)
+	if err := server.Run(); err != nil {
+		t.Fatal(err)
+	}
+	client := rpc.NewClient(listener.Addr().String())
+	defer func() {
+		client.Detach(true)
+	}()
+	fn(&FakeTerminal{t, client, DebugCommands(client), New(client, nil)})
+}
 
 func TestCommandDefault(t *testing.T) {
 	var (
@@ -103,4 +169,16 @@ func TestExecuteFile(t *testing.T) {
 func TestIssue354(t *testing.T) {
 	printStack([]api.Stackframe{}, "")
 	printStack([]api.Stackframe{{api.Location{PC: 0, File: "irrelevant.go", Line: 10, Function: nil}, nil, nil}}, "")
+}
+
+func TestIssue411(t *testing.T) {
+	withTestTerminal("math", t, func(term *FakeTerminal) {
+		term.MustExec("break math.go:8")
+		term.MustExec("trace math.go:9")
+		term.MustExec("continue")
+		out := term.MustExec("next")
+		if !strings.HasPrefix(out, "> main.main()") {
+			t.Fatalf("Wrong output for next: <%s>", out)
+		}
+	})
 }

